### PR TITLE
tomasdousek - ditcher

### PR DESCRIPTION
**Pull request by TomDousek**
# submission name:

tomasdousek
# upload zipped `archive.zip`

[archive.zip](https://raw.githubusercontent.com/TomDousek/TomDousek.github.io/main/media/archive.zip)

# report.txt

```
=== Evaluation config ===
  batch_size: 16
  device: None
  num_threads: 2
  prefetch_queue_depth: 4
  report: report.txt
  seed: 1234
  submission_dir: submissions/tomasdousek
  uncompressed_dir: videos
  video_names_file: public_test_video_names.txt
=== Evaluation results over 600 samples ===
  Average PoseNet Distortion: 0.00358563
  Average SegNet Distortion: 0.00328412
  Submission file size: 290,747 bytes
  Original uncompressed size: 37,545,463 bytes
  Compression Rate: 0.00774387
  Final score: 100*segnet_dist + √(10*posenet_dist) + 25*rate = 0.71
```
# does your submission require gpu for evaluation (inflation)?

yes
# did you include the compression script? and want it to be merged?

yes
# additional comments

**Visualizations have been moved to separate github page**
**[Click here to view the Interactive Page](https://tomdousek.github.io/)**

My initial idea was to use ffmpeg to compress the video as much as possible and then use learned upscaler that would enhance the video. This approach proved not very promising because ffmpeg with large -crf produced frames with large cubic artefacts that lacked texture which PoseNet needs. I was quite amazed when I saw Quatizr's first approach that you can extract the masks and poses and discard the original video completely (therefore name ditcher). 

My approach uses generator that extracts masks, inter-pair poses, and intra-pair poses, and generates pairs of yuv6 frames. The generator consists of a MaskEncoder that extracts features from mask, and a FrameHead that applies shifts to the features. They both use deptwise separated convolutions with residual connections, and SqueezeExcitation and PixelAttention blocks.

I hope you will like my visualizations as much as I do. I really liked the whole process of working on this challenge because the decompressed videos are visually very appealing and they often seemed like 80's video games with lots of neons. Thanks to the uncertainty video, I was able to find my model's weak spots and target them with special loss functions such as boundary loss. It displays also the original mask for refference, segnet and posenet losses and added speedometer for fun :). The uncertainty map clearly shows that biggest improvement in segenet distortion can be made in distant objects. 

The embedding space aptly shows relationships between classes. For example, the model can well separate undrivable from rest of objects, and interestingly, ego-vehicle and road lanes are closely related in the embedding space. You can also see that ego-vehicle and vehicle being on the other sides of embedding space, maybe this means that other cars are seen as obstacles from the model :). 

I have also added model_visualization video that shows what happens in each step in generator with the input data. The colors are PCA approximations into rgb space (similar colors means the convolution layers interpret the area as separate object). You can see, that most details are lost in the U-net down layer. If I had more time, this would be the place to focus on and try to conserve more details. I am also adding three screenshots from three distinct models that I tried to train along the way.

The visualizations can be found at https://tomdousek.github.io/

### CODE REVIEW

**Comment by YassineYousfi** [comment]
nice visualizations! can you host them somewhere else? I would like to keep the repo lightweight.

**Comment by TomDousek** [comment]
> nice visualizations! can you host them somewhere else? I would like to keep the repo lightweight.

Done!
The visualizations have been moved to https://tomdousek.github.io/

**Comment by TomDousek** [comment]
> nice visualizations! can you host them somewhere else? I would like to keep the repo lightweight.

I appreciate your feedback @YassineYousfi . In case you would like to read more about my solution, I have added more detailed description of my Generator network into my github page: https://tomdousek.github.io/
I am glad I put my work into this challenge learned a lot and did not vibe-code everything like the best submissions.

**Comment by YassineYousfi** [comment]
@TomDousek Thanks for the amazing write-up. We had an impressive batch of submissions, and after a lot of consideration, we decided to award the prize to Aaron's submission: https://aaronleslie.dev/blog/comma-compression

That said, we’d still love to link your write-up in the README if you’re okay with that.

**Comment by YassineYousfi** [comment]
@TomDousek Congratulations on making it onto the leaderboard! I hope you had fun competing in the challenge and learned something new along the way. If you are looking for a job or internship, please email givemeajob@comma.ai with a link to this PR. See you in the next challenge!

**Comment by TomDousek** [comment]
@YassineYousfi I absolutely agree. My jaw dropped, when I saw Aaron's work. I love the interactive animations! It reminds me of World Model paper from David Ha which I really enjoyed. 
I would be honored if you linked my write-up in the README. I hope it will shed light on what is going on inside the neural network to others.
Thank you so much for the opportunity. I am also working on the controlls challenge so I will then write you an email.
